File: infrastructure/spring-boot-r2dbc-persistence-adapter/src/main/java/codes/fdk/blueprint/api/infrastructure/persistence/r2dbc/CategoryEntityRepository.java
package codes.fdk.blueprint.api.infrastructure.persistence.r2dbc;

import codes.fdk.blueprint.api.domain.model.CategoryId;
import org.springframework.data.r2dbc.repository.R2dbcRepository;
import reactor.core.publisher.Flux;

//TODO auditing, profiling, projections?
interface CategoryEntityRepository extends R2dbcRepository<CategoryEntity, CategoryId> {
    Flux<CategoryEntity> findByParentId(CategoryId parentId);
}

File: infrastructure/vertx-postgres-persistence-adapter/src/main/kotlin/codes/fdk/blueprint/api/infrastructure/persistence/postgres/CategoryEntityRepository.kt
package codes.fdk.blueprint.api.infrastructure.persistence.postgres

import codes.fdk.blueprint.api.domain.model.CategoryId
import kotlinx.coroutines.flow.Flow

interface CategoryEntityRepository {

    companion object {
        const val TABLE_NAME = "categories"
    }

    suspend fun save(entity: CategoryEntity): CategoryEntity
    suspend fun findAll(): Flow<CategoryEntity>
    suspend fun findById(id: CategoryId): CategoryEntity?
    suspend fun findByParentId(parentId: CategoryId): Flow<CategoryEntity>

}